src/app/qgsmaptooltrimextendfeature.cpp

```diff
@@ -0,0 +1,283 @@
+/***************************************************************************
+    qgmaptoolextendfeature.cpp  -  map tool for extending feature
+    ---------------------
+    begin                : October 2018
+    copyright            : (C) 2018 by Loïc Bartoletti
+    email                : loic dot bartoletti at oslandia dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 3 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmaptooltrimextendfeature.h"
+
+#include "qgsmapcanvas.h"
+#include "qgsvertexmarker.h"
+#include "qgsvectorlayer.h"
+#include "qgsgeometry.h"
+#include "qgsrubberband.h"
+#include "qgssnappingutils.h"
+#include "qgstolerance.h"
+#include "qgisapp.h"
+#include "qgsgeometryutils.h"
+#include "qgsmapmouseevent.h"
+#include "qgssnapindicator.h"
+
+class FeatureFilter : public QgsPointLocator::MatchFilter
+{
+  public:
+    FeatureFilter()
+    {}
+
+    bool acceptMatch( const QgsPointLocator::Match &match ) override
+    {
+      if ( mLayer )
+        return match.layer() == mLayer && match.hasEdge();
+
+      return match.hasEdge();
+    }
+    // We only want to modify the current layer. When geometries are overlapped, this makes it possible to snap onto the current layer.
+    void setLayer( QgsVectorLayer *vlayer ) { mLayer = vlayer; }
+
+  private:
+    const QgsVectorLayer *mLayer = nullptr;
+};
+
+QgsMapToolTrimExtendFeature::QgsMapToolTrimExtendFeature( QgsMapCanvas *canvas )
+  : QgsMapToolEdit( canvas )
+{
+  mToolName = tr( "Trim/Extend feature" );
+}
+
+QgsMapToolTrimExtendFeature::~QgsMapToolTrimExtendFeature()
+{
+}
+
+static void getPoints( const QgsPointLocator::Match &match, QgsPoint &p1, QgsPoint &p2 )
+{
+  const QgsFeatureId fid = match.featureId();
+  const int vertex = match.vertexIndex();
+
+  const QgsGeometry geom = match.layer()->getGeometry( fid );
+
+  if ( !( geom.isNull() || geom.isEmpty() ) )
+  {
+    p1 = geom.vertexAt( vertex );
+    p2 = geom.vertexAt( vertex + 1 );
+  }
+}
+
+void QgsMapToolTrimExtendFeature::canvasMoveEvent( QgsMapMouseEvent *e )
+{
+  mapPoint = e->mapPoint();
+
+  FeatureFilter filter;
+  QgsPointLocator::Match match;
+
+  switch ( step )
```

Comment:
Do you think that `step` could be an enum? It would be clearer.

Reply:
Done